internal/goro/group.go

```diff
@@ -0,0 +1,77 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package goro
+
+import (
+	"context"
+	"sync"
+)
+
+// Group manages a set of long-running goroutines as a unit. Group differs from
+// `errgroup.Group` in that it does not link the lifetimes of the spawned
+// goroutines and it is not expected that a client will call Wait() without
+// first calling Cancel(). If you're looking for a short-lived (e.g.
+// request-scoped) group of transient goroutines, you probably want
+// `errgroup.Group`. The expectation is that an instance of this type will be
+// held as a member field of an object that has standard Start()/Stop()
+// semantics (e.g. a `common.Daemon`). The zero-value of this type is valid.
+type Group struct {
+	initOnce sync.Once
+	ctx      context.Context
+	cancel   context.CancelFunc
+	wg       sync.WaitGroup
+}
+
+// Go spawns a goroutine whose lifecycle will be managed via this Group object.
+// All functions passed to Go on the same instance of Group will be given the
+// same `context.Context` object, which will be used to indicate cancellation.
+// If the supplied func does not abide by `ctx.Done()` of the provided
+// `context.Context` then `Wait()` on this `Group` will hang.
```

Comment:
```suggestion
// `context.Context` then `Wait()` on this `Group` will hang until function exits.
```
or something like this.